### DISCUSSION TITLE
Custom Exceptions without mapping logic

### DISCUSSION THREAD

**Discussion by ghost** [mapstruct/mapstruct#3047] [Q&A]
I feel like I'm either doing something wrong, or going about it in the wrong way.

I have a simple situation:

```java
@ValueMapping(source = "UNSPECIFIED", target = MappingConstants.THROW_EXCEPTION)
@ValueMapping(source = "UNRECOGNIZED", target = MappingConstants.THROW_EXCEPTION)
Target toEntity(Source grpc);
```

It doesn't have to be a value mapping, the reference guide doesn't seem to suggest any way of just doing a custom exception wrapper. Unless I'm missing something.

I want to make a custom exception, so that I can give extra information on what is happening. The default exception is okay, I would like to wrap it.

But unlike the [10.12 #Exceptions](https://mapstruct.org/documentation/stable/reference/html/#exceptions), I don't want to do any mapping by hand. It sorta defeats the purpose. I'd rather just do the whole mapping by hand at that point.

I get what the implementation does. It lets me define a throws clause, so that I can do a manual mapping and not automatically wrap my hand-made mapping and subsequent exception in a `RuntimeException`. I get that, and it's not what I would ideally want to begin with.

I would to either do the following:

Method 1:

```java
@Mapper(... uses = TheAboveMapper.class)
@RequiredArgsConstructor
public abstract class OutterMapper {
    
    private final TheAboveMapper aboveMapper;
    
    public Target toEntity(Source grpc) {
        try {
            return aboveMapper.toEntity(source);
        } catch(IllegalArgumentException e) {
            throw new MyWrappedException("Details", e);
        }
    }
}
```

Method 2:

```java
@BeforeMapping
protected void validate(Source source) {
    // test, throw exception if needed
}
```

Method 1, not a big fan of. The try catch block is too big, other things could go wrong, not sure I'll catch the exception I think I'll catch.

Method 2, I think will have `MyException` be wrapped in a RuntimeException. Maybe I can avoid this with `throws MyException` on the mapping definition. But even if it works ... I would like to reuse that code. How could I reuse that `@BeforeMapping`? Could I add it to a config, or an annotation? Or at the very least, with type matching?

I would rather not type match, since I could do a generic impleementation that is Enum `.class` independent and would only use the names of the fields. 

Or am I really going the wrong way on this?

**Comment by filiphr** [comment]
I am not sure that I entirely understand what you are trying to achieve. Are you trying to perform some kind of validation on the source and propagate that exception?

There are different options that you can do. First thing is that MapStruct doesn't usually generate exception throwing, at least not in the normal path of mapping. For some conversions it does do this, but this is for those out-of-the-box conversions that have some kind of checked exception (usually around XML types).

Another place where MapStruct throws an exception is in the value mapping. For the value mapping you can actually use `EnumMapping#unexpectedValueMappingException` or `Mapper#unexpectedValueMappingException` or `MapperConfig#unexpectedValueMappingException`. That would throw that exception instead of the `IllegalArgumentException` for value mappings. MapStruct will only wrap that exception in a runtime exception if the exception is typed. Otherwise it will not do anything.

Wrapping the exception is done only for checked exceptions that are not thrown by the mapping method itself.

If you are talking about some validation logic and / or want to reuse certain things you could use the concept of [Before / After Mapping](https://mapstruct.org/documentation/stable/reference/html/#customizing-mappings-with-before-and-after) and `Mapper#uses`. Basically, with `Mapper#uses` you are instructing for MapStruct to look for other lifecycle, mapping or other methods.

e.g.

```
public class ValidationUtils {

    @BeforeMapping
    public static void validateSource(Source source) {
        // perform your validations
    }

}
```

Note: the `validateSource` method doesn't have to be static, it can be a normal method (MapStruct will then instantiate the `ValidationUtils` or inject it if you are using some DI framework). 

As I mentioned earlier if your `MyException` is a checked exception and your mapping method does not defined `throws MyException` then MapStruct will wrap it in an `RuntimeException`. However, if your exception is a runtime exception then MapStruct won't do anything, because you do not need to define in `validateSource` that you are throwing that.

Hope this helps.
